File: C/21_Binary_Search/BOJ_05_2110_공유기설치.c
#include <stdio.h>
#include <stdlib.h>
int compare(const void* a, const void* b) {
	int o1 = *(int*)a;
	int o2 = *(int*)b;
	return o1 - o2;
}
main() {
	int n, m;
	scanf("%d %d", &n, &m);
	int* arr = (int*)malloc(sizeof(int) * n);
	for (int i = 0; i < n; i++) {
		scanf("%d", &arr[i]);
	}
	qsort(arr, n, sizeof(int), compare);
	int max = arr[n - 1] - arr[0], min = 0;
	while (min <= max) {
		int mid = (max + min) / 2;
		int cnt = 1, start = arr[0];
		for (int i = 1; i < n; i++) {
			if (arr[i] - start >= mid) {
				cnt++;
				start = arr[i];
			}
		}
		if (cnt < m) {
			max = mid - 1;
		}
		else {
			min = mid + 1;
		}
	}
	printf("%d", max);
}

File: JAVA/21_Binary_Search/BOJ_05_2110_공유기설치.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine(), " ");
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int[] arr = new int[n];
        for (int i = 0; i < n; i++) {
            arr[i] = Integer.parseInt(br.readLine());
        }
        Arrays.sort(arr);
        int max = arr[n - 1] - arr[0], min = 1;
        while (min <= max) {
            int mid = (max + min) / 2;
            int cnt = 1, start = arr[0];
            for (int i = 1; i < n; i++) {
                if (arr[i] - start >= mid) {
                    cnt++;
                    start = arr[i];
                }
            }
            if (cnt < m) {
                max = mid - 1;
            } else {
                min = mid + 1;
            }
        }
        System.out.println(max);
    }
}